client/client.go

```diff
@@ -188,11 +191,11 @@ func (cli *Client) Close() error {
 // It appends the query parameters to the path if they are not empty.
 func (cli *Client) getAPIPath(ctx context.Context, p string, query url.Values) string {
 	var apiPath string
-	if cli.negotiateVersion && !cli.negotiated {
+	if cli.negotiateVersion && !cli.isVersionNegotiated() {
 		cli.NegotiateAPIVersion(ctx)
```

Comment:
There is nothing preventing the client from calling `NegotiateAPIVersion` concurrently from multiple goroutines, so an unbounded number of duplicate ping requests could be in-flight at once. A client only needs one version negotiation to be in-flight at a time; any more only serves to increase the API request load on the daemon.

Reply:
I'll check to be sure, yet expect this to be a sufficient fix unless I miss something; in my case of the `whalewatcher` OpenSource module there is a "safe" location before kicking off the separate event-streaming and processing goroutines and as the race detector seems to triggered exactly when using this module I assume that running an explicit `NegotiateAPIVersion` should do the trick and additionally avoiding spamming the Docker daemon.

Reply:
@crigor 👍🏼 fixes my issues for the moment. Thank you very much for the help!

Reply:
@corhere we could also use a single flight for that I guess? https://github.com/thaJeztah/docker/commit/fac9fc242abec4659c6074cf52c89057b5bc7f52

Reply:
Actually wondering now if a `sync.Once` would do the job (will have to check the logic on what we currently do on a failed negotiation)

Reply:
The error return from `cli.Ping` is ignored...
https://github.com/moby/moby/blob/a347f7971110b0eef2cd644327711b3cf946d52a/client/client.go#L223-L228
...so a failed request or canceled context will result in the version being "negotiated" to `1.24` for all subsequent requests, or until the consumer explicitly calls `NegotiateAPIRequest` or `NegotiateAPIVersionPing`.
https://github.com/moby/moby/blob/a347f7971110b0eef2cd644327711b3cf946d52a/client/client.go#L249-L272
That doesn't sound like desirable behaviour; having a canceled context for one request negatively effect subsequent contexts. Anyway, I am pretty sure replacing the `negotiated` flag with a `sync.Once` would achieve the same effective semantics in a concurrent world while suppressing redundant pings.
```go
func (cli *Client) getAPIPath(ctx context.Context, p string, query url.Values) string {
	if cli.negotiateVersion {
		cli.negotiateOnce.Do(func() { cli.NegotiateAPIVersion(ctx) })
	}
	/* ... */
}
```